### PR TITLE
Retry upload alert

### PR DESCRIPTION
**Pull request by lorenzoPrimi**
In `UploadFooterViewController.h` not the `uploadMeasurements` method gets an array of measurements and the index where to start from.
If an upload fails and the user press Retry, the index will be the current failed upload measurement.

The change in the `Result.h` file in the NOT_UPLOADED_QUERY is due to the face we don't want to 
show the not uploaded banner if there are failed or re_run tests not uploaded.

### CODE REVIEW

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 1]
```diff
@@ -65,45 +65,46 @@ -(void)uploadResult{
...
-        [self uploadSingleMeasurement:self.measurement];
-        dispatch_async(dispatch_get_main_queue(), ^{
-            [MBProgressHUD HUDForView:self.navigationController.view].progress = 1;
-        });
+        [self uploadMeasurements:[NSArray arrayWithObject:self.measurement]];
     }
 }
 
--(void)uploadMeasurements:(SRKResultSet *)notUploaded{
+-(void)uploadMeasurements:(NSArray *)notUploaded {
```
Comment (lorenzoPrimi) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: `SRKResultSet` is a subclass of `NSArray`

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 2]
```diff
@@ -65,45 +65,46 @@ -(void)uploadResult{
...
+    [self uploadMeasurements:notUploaded startAt:0];
+}
+
+-(void)uploadMeasurements:(NSArray *)notUploaded startAt:(int)idx{
     if ([notUploaded count] == 0) return;
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Please, use `<= 0` here to also cover the unlikely but possible case where the index is negative.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 3]
```diff
@@ -65,45 +65,46 @@ -(void)uploadResult{
...
-            [MBProgressHUD HUDForView:self.navigationController.view].label.text = NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading", [NSString stringWithFormat:@"%d/%ld", done, [notUploaded count]]);
-        });
-        [self uploadSingleMeasurement:currentMeasurement];
-        progress += measurementValue;
-        done++;
+    while (idx < [notUploaded count]){
+        Measurement *currentMeasurement = [notUploaded objectAtIndex:idx];
         dispatch_async(dispatch_get_main_queue(), ^{
-            [MBProgressHUD HUDForView:self.navigationController.view].progress = progress;
+            [MBProgressHUD HUDForView:self.navigationController.view].label.text = NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading", [NSString stringWithFormat:@"%d/%ld", idx+1, [notUploaded count]]);
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Can you break the line after `NSLocalizedFormatString` to increase the diff readability?

File: ooniprobe/Model/Database/Result.h [hunk 4]
```diff
@@ -2,7 +2,7 @@
 #import <SharkORM/SharkORM.h>
 #import "JsonResult.h"
 #import "Measurement.h"
-#define NOT_UPLOADED_QUERY @"is_failed = 0 AND (is_uploaded = 0 || report_id IS NULL)"
+#define NOT_UPLOADED_QUERY @"is_failed = 0 AND is_rerun = 0 AND is_done = 1 AND (is_uploaded = 0 || report_id IS NULL)"
```
Comment (bassosimone) [inline ooniprobe/Model/Database/Result.h]: Did you add these checks because the underlying set could mutate as we process it?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 1 above)
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Since this may also surprise others in the future, maybe add a comment in the code?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 5]
```diff
@@ -118,6 +119,20 @@ -(void)uploadSingleMeasurement:(Measurement*)measurement{
         [measurement setReport_id:[results updatedReportID]];
         [measurement save];
     }
+    return [results good];
 }
 
+-(void)showRetryPopup:(NSArray *)notUploaded :(int)start{
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Here it would be better to remove the space between `notUploaded` and `:`

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 6]
```diff
@@ -118,6 +119,20 @@ -(void)uploadSingleMeasurement:(Measurement*)measurement{
...
+    UIAlertAction* okButton = [UIAlertAction
+                               actionWithTitle:NSLocalizedString(@"Modal.Retry", nil)
+                               style:UIAlertActionStyleDefault
+                               handler:^(UIAlertAction * action) {
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I'd remove the space between `*` and `action`

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 2 above)
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: An NSArray count can't be negative by implementation. It return an unsigned NSInteger

File: ooniprobe/Model/Database/Result.h (hunk 4 above)
Comment (lorenzoPrimi) [reply ooniprobe/Model/Database/Result.h]: No, just because we don't care about re_run tests and of course the unfinished ones (they will never show up in results)

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 5 above)
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: It's obj-c syntax. I can't remove the space, I can only add a name for the param, will do that.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 2 above)
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Gotcha

File: ooniprobe/Model/Database/Result.h (hunk 4 above)
Comment (bassosimone) [reply ooniprobe/Model/Database/Result.h]: Gotcha

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 5 above)
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: ah, wow, I completely missed this aspect. Yes, please!

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 7]
```diff
@@ -65,45 +65,48 @@ -(void)uploadResult{
...
-        });
-        [self uploadSingleMeasurement:currentMeasurement];
-        progress += measurementValue;
-        done++;
+    while (idx < [notUploaded count]){
+        Measurement *currentMeasurement = [notUploaded objectAtIndex:idx];
         dispatch_async(dispatch_get_main_queue(), ^{
-            [MBProgressHUD HUDForView:self.navigationController.view].progress = progress;
+            [MBProgressHUD HUDForView:self.navigationController.view].label.text =
+            NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading", [NSString stringWithFormat:@"%d/%ld", idx+1, [notUploaded count]]);
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I meant to use this formatting, which is less suprising. Please apply it:

```ObjectiveC
[MBProgressHUD HUDForView:self.navigationController.view].label.text = NSLocalizedFormatString(
    @"Modal.ResultsNotUploaded.Uploading",
    [NSString stringWithFormat:@"%d/%ld", idx+1, [notUploaded count]]);
```

Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I don't like to cut after NSLocalizedFormatString(
any other solution?

Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]:
```ObjectiveC
[MBProgressHUD HUDForView:self.navigationController.view].label.text =
    NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading",
        [NSString stringWithFormat:@"%d/%ld", idx+1, [notUploaded count]]);
```

?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 8]
```diff
@@ -90,7 +90,8 @@ -(void)uploadMeasurements:(NSArray *)notUploaded startAt:(int)idx{
         Measurement *currentMeasurement = [notUploaded objectAtIndex:idx];
         dispatch_async(dispatch_get_main_queue(), ^{
             [MBProgressHUD HUDForView:self.navigationController.view].label.text =
-            NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading", [NSString stringWithFormat:@"%d/%ld", idx+1, [notUploaded count]]);
+            NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading",
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Please also indent `NSLocalizedFormatString` four spaces on the right, otherwise it's not apparent that this line is a continuation from the previous line.

File: OONIProbeUnitTests/ResultTest.m [hunk 9]
```diff
@@ -32,17 +32,18 @@ - (void)testNoMeasurements {
 }
 
 - (void)testNotUploaded {
-    //One measurement, isEveryMeasurementUploaded false
+    //One measurement, with is_done false, isEveryMeasurementUploaded true
     Result *result = [Result new];
-    Measurement *measurement = [self addMeasurement:result];
-    XCTAssert(![result isEveryMeasurementUploaded]);
+    [self addMeasurement:result];
+    XCTAssert([result isEveryMeasurementUploaded]);
```
Comment (bassosimone) [inline OONIProbeUnitTests/ResultTest.m]: I do not get exactly why the code change broke the original test. Can you explain?

File: OONIProbeUnitTests/ResultTest.m [hunk 10]
```diff
@@ -32,17 +32,18 @@ - (void)testNoMeasurements {
...
 
 - (void)testUploaded {
     //Measurement uploaded but report_id null, isEveryMeasurementUploaded false
     Result *result = [Result new];
     Measurement *measurement = [self addMeasurement:result];
     measurement.is_uploaded = true;
+    measurement.is_done = true;
     [measurement save];
     XCTAssert(![result isEveryMeasurementUploaded]);
 }
```
Comment (bassosimone) [inline OONIProbeUnitTests/ResultTest.m]: Can you also explain why here the test needed to be changed?

File: ooniprobe/Model/Database/Result.h (hunk 4 above)
Comment (bassosimone) [reply ooniprobe/Model/Database/Result.h]: One more question: why don't we care about rerun tests?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 11]
```diff
@@ -65,45 +65,49 @@ -(void)uploadResult{
...
     float measurementValue = 1.0/[notUploaded count];
-    int done = 1;
-    for (Measurement *currentMeasurement in notUploaded){
-        dispatch_async(dispatch_get_main_queue(), ^{
-            [MBProgressHUD HUDForView:self.navigationController.view].label.text = NSLocalizedFormatString(@"Modal.ResultsNotUploaded.Uploading", [NSString stringWithFormat:@"%d/%ld", done, [notUploaded count]]);
-        });
-        [self uploadSingleMeasurement:currentMeasurement];
-        progress += measurementValue;
-        done++;
+    while (idx < [notUploaded count]){
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I think the type of `idx` should be the same type returnd by `[notUploaded count]`

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m [hunk 12]
```diff
@@ -65,45 +65,49 @@ -(void)uploadResult{
...
+            progress += measurementValue;
+            idx++;
+            dispatch_async(dispatch_get_main_queue(), ^{
+                [MBProgressHUD HUDForView:self.navigationController.view].progress = progress;
+            });
+        }
+        else {
+            [self showRetryPopup:notUploaded startAt:idx];
+            return;
+        }
```
Comment (bassosimone) [inline ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I believe the complexity of the function would be reduced if you'd write:

```ObjectiveC
  if (![self uploadMeasurement:currentMeasurement]) {
    [self showRetryPopup:notUploaded startAt:idx];
    return;
  }
 // other code
```

File: ooniprobe/Model/Database/Result.h (hunk 4 above)
Comment (lorenzoPrimi) [reply ooniprobe/Model/Database/Result.h]: because a test who is re ran is usually a failed test and it has no json.
Furthermore is not visible in the list so it wil never get the upload bottom banner

Comment (bassosimone) [reply ooniprobe/Model/Database/Result.h]: Right. Is it correct to say, though, that in principle a re-run test may not be failed? What happens in that case? Is the output file overwritten? Is this a corner case we need to care (maybe just by opening an issue and dealing with it in the future)?

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 11 above)
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: it is the same, they are both int

File: OONIProbeUnitTests/ResultTest.m (hunk 10 above)
Comment (lorenzoPrimi) [reply OONIProbeUnitTests/ResultTest.m]: because now we don't care about tests who are not done. 
This is how the test should have been written since the beginning

File: OONIProbeUnitTests/ResultTest.m (hunk 9 above)
Comment (lorenzoPrimi) [reply OONIProbeUnitTests/ResultTest.m]: a test who is not done can't be reuploaded

File: ooniprobe/Model/Database/Result.h (hunk 4 above)
Comment (lorenzoPrimi) [reply ooniprobe/Model/Database/Result.h]: a test who has been re ran is failed.
Everything will be overwritten in the moment it will be re run, except the db row (a new one is created)

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 11 above)
Comment (bassosimone) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: Are you sure? NSUInteger strikes me as an _unsigned_ integer:

https://developer.apple.com/documentation/foundation/nsarray/1409982-count?language=objc

File: ooniprobe/Model/Database/Result.h (hunk 4 above)
Comment (bassosimone) [reply ooniprobe/Model/Database/Result.h]: Right, we can only re-run a test that failed. We may offer used to re-run a test in general in the future. When we will do that, we will probably need to think at the implications.

File: ooniprobe/View/TestResults/Footer/UploadFooterViewController.m (hunk 11 above)
Comment (lorenzoPrimi) [reply ooniprobe/View/TestResults/Footer/UploadFooterViewController.m]: I don't really know good these obj-c types but surely I will rename it

File: OONIProbeUnitTests/ResultTest.m (hunk 9 above)
Comment (bassosimone) [reply OONIProbeUnitTests/ResultTest.m]: So, the function should probably be renamed `testNotDone`?

File: OONIProbeUnitTests/ResultTest.m (hunk 10 above)
Comment (bassosimone) [reply OONIProbeUnitTests/ResultTest.m]: ack
